common/dev/peci.c

```diff
@@ -0,0 +1,268 @@
+				cfg->cache_status = SENSOR_FAIL_TO_ACCESS;
+				break;
+			case PECI_CC_ILLEGAL_REQUEST:
+				printf("PECI sensor [%x] unknown request\n", sensor_num);
+				cfg->cache_status = SENSOR_NOT_FOUND;
+				break;
+			default:
+				printf("PECI sensor [%x] control hardware, firmware or associated logic error\n",
+				       sensor_num);
+				cfg->cache_status = SENSOR_UNSPECIFIED_ERROR;
+				break;
+			}
+		}
+
+		free(read_buf);
+		return false;
+	}
+
+	complete_code = &read_buf[PECI_READ_LEN];
+	retry_count = PECI_RETRY_MAX;
+	while (1) {
+		ret = peci_read(cmd, cfg->port, index[1], param[1], PECI_READ_LEN,
+				&read_buf[PECI_READ_LEN]);
+		if (ret == 0) {
+			if (*complete_code == PECI_CC_RSP_SUCCESS) {
+				break;
+			}
+		} else if (--retry_count) { // wait for 10ms before next retry
+			k_msleep(10);
+		} else { // all retries failed exit the loop
+			break;
+		}
+	}
+	if (retry_count == 0) {
+		if (ret) {
+			printf("PECI sensor [0x%x] retry read fail, retry time: %d, completion_code: 0x%x\n",
+			       sensor_num, PECI_RETRY_MAX, *complete_code);
+			cfg->cache_status = SENSOR_FAIL_TO_ACCESS;
+		} else {
+			switch (*complete_code) {
+			case PECI_CC_RSP_TIMEOUT:
+			case PECI_CC_OUT_OF_RESOURCES_TIMEOUT:
+				printf("PECI sensor [0x%x] response timeout\n", sensor_num);
+				cfg->cache_status = SENSOR_FAIL_TO_ACCESS;
+				break;
+			case PECI_CC_ILLEGAL_REQUEST:
+				printf("PECI sensor [0x%x] unknown request\n", sensor_num);
+				cfg->cache_status = SENSOR_NOT_FOUND;
+				break;
+			default:
+				printf("PECI sensor [0x%x] control hardware, firmware or associated logic error\n",
+				       sensor_num);
+				cfg->cache_status = SENSOR_UNSPECIFIED_ERROR;
+				break;
+			}
+		}
+
+		free(read_buf);
+		return false;
+	}
+
+	pkg_energy = read_buf[4];
+	pkg_energy = (pkg_energy << 8) | read_buf[3];
+	pkg_energy = (pkg_energy << 8) | read_buf[2];
+	pkg_energy = (pkg_energy << 8) | read_buf[1];
+
+	run_time = read_buf[9];
+	run_time = (run_time << 8) | read_buf[8];
+	run_time = (run_time << 8) | read_buf[7];
+	run_time = (run_time << 8) | read_buf[6];
+
+	// first read, need second data to calculate
+	if ((last_pkg_energy == 0) && (last_run_time == 0)) {
+		last_pkg_energy = pkg_energy;
+		last_run_time = run_time;
+		free(read_buf);
+		return false;
+	}
+
+	if (pkg_energy >= last_pkg_energy) {
+		diff_energy = pkg_energy - last_pkg_energy;
+	} else {
+		diff_energy = pkg_energy + (0xffffffff - last_pkg_energy + 1);
+	}
+	last_pkg_energy = pkg_energy;
+
+	if (run_time >= last_run_time) {
+		diff_time = run_time - last_run_time;
+	} else {
+		diff_time = run_time + (0xffffffff - last_run_time + 1);
+	}
+	last_run_time = run_time;
+
+	if (diff_time == 0) {
+		free(read_buf);
+		return false;
+	} else {
+		// energy / unit time
+		*reading = ((float)diff_energy / (float)diff_time * 0.06103515625);
+		free(read_buf);
+		return true;
+	}
+}
+
+bool peci_sensor_read(uint8_t sensor_num, float *reading)
```

Comment:
check reading != NULL

Reply:
The optimization code is based on v2022.07.01. After that, the main branch has been phased in some
commits related to sensor monitoring. Those changes will be phased in main_v2 in the other PRs. This
problem will be solved in those PRs.
Please check #223  .